### DISCUSSION TITLE
用 wsl2 运行 cli 7.6.7 版本的 docker-easyconnect 主机连接失败

### DISCUSSION THREAD

**Discussion by String10** [docker-easyconnect/docker-easyconnect#70] [Q&A] [answered]
我用的是 win 版的 Docker ，在 wsl2 中运行指令：
```shell
docker run --device /dev/net/tun --cap-add NET_ADMIN -ti -p 127.0.0.1:8888:8888 -e EC_VER=7.6.7 -e CLI_OPTS="-d vpnaddress -u username -p password" hagb/docker-easyconnect:cli
```
出现下面的信息：
```shell
iptables v1.8.2 (nf_tables): Chain 'MASQUERADE' does not exist
Try `iptables -h' or 'iptables --help' for more information.
iptables v1.8.2 (nf_tables):  RULE_INSERT failed (No such file or directory): rule in chain INPUT
iptables v1.8.2 (nf_tables):  RULE_INSERT failed (No such file or directory): rule in chain INPUT
iptables v1.8.2 (nf_tables):  RULE_INSERT failed (No such file or directory): rule in chain INPUT
iptables v1.8.2 (nf_tables): Couldn't load match `state':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
auto login is disabled
auth failed, please check if the vpn address is valid and reachable!
login failed!
svpn stop!
auto login is disabled
user "*****" login successfully!
```

然后在主机浏览器上设置 socks5 代理，按照步骤设置使用 127.0.0.1:1080，但是还是失败。

我尝试了切换 iptables 的版本，但是还是有上面的报错信息。

**Comment by Hagb** [comment, marked as answer]
> 我尝试了切换 iptables 的版本，但是还是有上面的报错信息。

切换了是指加了 `-e IPTABLES_LEGACY=1` 吗？能否把运行的命令行和输出也发上来？

**Reply by String10** [reply]
我之前使用的是```update-alternatives --config iptables```来进行切换的，
加了```-e IPTABLES_LEGACY=1```之后运行截图见下图：
![image](https://user-images.githubusercontent.com/69832835/130164582-29c69eec-1d58-47d0-9640-1bb6e42ba91b.png)
具体输出为：
```
update-alternatives: using /usr/sbin/iptables-legacy to provide /usr/sbin/iptables (iptables) in manual mode
update-alternatives: using /usr/sbin/ip6tables-legacy to provide /usr/sbin/ip6tables (ip6tables) in manual mode
auto login is disabled
auth failed, please check if the vpn address is valid and reachable!
login failed!
svpn stop!
auto login is disabled
user "******" login successfully!
```
浏览器 socks5 具体设置为：
![![image](https://user-images.githubusercontent.com/69832835/130164582-29c69eec-1d58-47d0-9640-1bb6e42ba91b.png)](https://user-images.githubusercontent.com/69832835/130164578-2ee7f1f8-1e86-4dad-a39a-dce8b6a5fcaf.png)

**Reply by Hagb** [reply]
> 我之前使用的是`update-alternatives --config iptables`来进行切换的，
> 加了`-e IPTABLES_LEGACY=1`之后运行截图见下图：
> ![image](https://user-images.githubusercontent.com/69832835/130164582-29c69eec-1d58-47d0-9640-1bb6e42ba91b.png)
> 具体输出为：
> 
> ```
> update-alternatives: using /usr/sbin/iptables-legacy to provide /usr/sbin/iptables (iptables) in manual mode
> update-alternatives: using /usr/sbin/ip6tables-legacy to provide /usr/sbin/ip6tables (ip6tables) in manual mode
> auto login is disabled
> auth failed, please check if the vpn address is valid and reachable!
> login failed!
> svpn stop!
> auto login is disabled
> user "******" login successfully!
> ```
> 
> 浏览器 socks5 具体设置为：
> ![image](https://user-images.githubusercontent.com/69832835/130164578-2ee7f1f8-1e86-4dad-a39a-dce8b6a5fcaf.png)

登录至少 10 秒后，运行 `docker exec 容器名 busybox ifconfig` 和 `time telnet 127.0.0.1 1080` 各输出什么？

浏览器那儿访问内网地址提示何种错误？

**Reply by Hagb** [reply]
1. 刚刚看了看 wsl 的文档，windows 似乎不能直接通过 `127.0.0.1` 来访问 wsl，详细请参看 https://docs.microsoft.com/zh-cn/windows/wsl/compare-versions#accessing-network-applications 。
2. 刚才才看到，你的参数里面开放的端口是 8888 （-e 127.0.0.1:8888:8888）而不是 1080。请把 `-e 127.0.0.1:8888:8888` 替换成 `-e 1080:1080` 或将后者加到参数中。

如果依然不行，尝试加上 `-e IPTABLES_LEGACY=1` 参数并提供上一条回复中所说的信息

**Reply by String10** [reply]
似乎将```8888:8888```替换为```1080:1080```之后连接成功了，感谢🙇‍~

**Comment by vzcrab** [comment]
能问下最后的命令是什么吗，同样的问题，日志一直在输出以下内容
```
user "***" auto login successfully
svpn stop!
```

**Reply by String10** [reply]
使用的命令即第一条中给出的，`user auto login successfully`应该是最后一条日志，像您这种情况最后输出了`svpn stop!`感觉不像是和我一样的问题。

**Reply by String10** [reply]
而且如果`svpn stop!`的话，在我这边会自动重连直到连接成功的。

**Reply by vzcrab** [reply]
貌似是我们学校 vpn 的问题

**Comment by jwqian99** [comment]
7.6.8 cli, windows 10 21H2, wsl2 docker. 命令中加入-e IPTABLES_LEGACY=1后容器正常连接至vpn。用clash能成功转发流量至socks5 127.0.0.1 port 1080。唯一不足就是延迟稍高。
